openmls/src/binary_tree/array_representation/tree.rs

```diff
@@ -0,0 +1,165 @@
+    /// be empty and has to yield a full, left-balanced binary tree. The nodes
+    /// in the tree are ordered in the array-representation. This function
+    /// throws a `InvalidNumberOfNodes` error if the number of nodes does not
+    /// allow the creation of a full, left-balanced binary tree and an
+    /// `OutOfRange` error if the number of given nodes exceeds the range of
+    /// `NodeIndex`.
+    pub(crate) fn new(nodes: Vec<T>) -> Result<Self, ABinaryTreeError> {
+        if nodes.len() > NodeIndex::max_value() as usize {
+            return Err(ABinaryTreeError::OutOfRange);
+        } else if nodes.len() % 2 != 1 {
+            return Err(ABinaryTreeError::InvalidNumberOfNodes);
+        }
+        Ok(ABinaryTree { nodes })
+    }
+
+    /// Obtain a reference to the data contained in the `Node` at index
+    /// `node_index`, where the indexing corresponds to the array representation
+    /// of the underlying binary tree. Returns `None` if the index is larger
+    /// than the size of the tree.
+    pub(in crate::binary_tree) fn node_by_index(
+        &self,
+        node_index: NodeIndex,
+    ) -> std::option::Option<&T> {
+        self.nodes.get(node_index as usize)
+    }
+
+    /// Return the number of nodes in the tree.
+    pub(in crate::binary_tree) fn size(&self) -> NodeIndex {
+        let len = self.nodes.len();
+        debug_assert!(len <= u32::MAX as usize);
+        self.nodes.len() as u32
+    }
+
+    /// Return the number of leaves in the tree.
+    pub(crate) fn leaf_count(&self) -> TreeSize {
+        (self.size() + 1) / 2
+    }
+
+    /// Return a vector of leaves sorted according to their position in the tree
+    /// from left to right. This function should not fail and only returns a
+    /// `Result`, because it might throw a
+    /// [LibraryError](ABinaryTreeError::LibraryError).
+    pub(crate) fn leaves(&self) -> Result<Vec<&T>, ABinaryTreeError> {
+        let mut leaf_references = Vec::new();
+        for leaf_index in 0..self.leaf_count() {
+            let node_index = usize::try_from(to_node_index(leaf_index))
+                // The tree size and thus the leaf count should fit into usize on 32
+                // bit architectures.
+                .map_err(|_| ABinaryTreeError::LibraryError)?;
+            let node_ref = self
+                .nodes
+                .get(node_index)
+                // Since the index is within the bounds of the tree, this should
+                // be Some.
+                .ok_or(ABinaryTreeError::LibraryError)?;
+            leaf_references.push(node_ref);
+        }
+        Ok(leaf_references)
+    }
+
+    /// Creates and returns an empty `AbDiff`.
+    pub(crate) fn empty_diff(&self) -> AbDiff<'_, T> {
+        self.into()
+    }
+
+    /// Merges the changes applied to the `StagedAbDiff` into the tree. This
+    /// function should not fail and only returns a `Result`, because it might
+    /// throw a [LibraryError](ABinaryTreeError::LibraryError).
```

Comment:
Please use something like 

```suggestion
/// throw a [`LibraryError`].
```

wherever possible to keep the comments more readable in code.